TERRAFORM.md

```diff
@@ -0,0 +1,96 @@
+# Terraform Best Practices and Guidelines
+
+General terraform guidelines and best practices
+
+## Table of Contents
+
+- [Table of Contents](#table-of-contents)
+- [Introduction](#introduction)
+- [Code Organization](#code-organization)
+  - [Reusable Module Usage](#reusable-module-usage)
+- [State Management](#state-management)
+  - [Remote State](#remote-state)
+    - [State key](#state-key)
+  - [State Locking](#state-locking)
+  - [State migration](#state-migration)
+  - [Module Guidelines](#module-guidelines)
+- [Code formatting](#code-formatting)
+
+## Introduction
+
+Terraform is a powerful tool for provisioning and managing infrastructure as code. To ensure maintainability, scalability, and security, it's important to follow best practices and guidelines when working with Terraform.
+
+## Code Organization
+
+The high level code organization should look like this
+
+```text
+|- modules/                   # terraform reusable modules for the whole repository
+|- ORGANIZATIONS_OR_ACCOUNT   # e.g. AWS Account name
+|           |- modules        # terraform reusable modules for this organization/account
+|           |- shared         # shared resources
+|           |- STAGING        # resource pro staging
+```
+
+Then it should mirror the hierarchical structure of the infrastructure. Note that each terraform module (project) should be kept small.
+
+### Reusable Module Usage
+
+Use modules to encapsulate and reuse configurations. This promotes DRY (Don't Repeat Yourself) principles and makes your configurations more manageable.
```

Comment:
We probably want to point out the caveats of modules here. The main one I think is that when you change a module, we don't currently have a good way to check what changes this triggers across the infrastructure and what needs to be pushed/applied.

Reply:
Good idea.

Actually we do have a way to check, we have our `tf-all` zsh function that gives you the possiblity to run `terraform plan` over the whole repository so you can check the whole diff in one command and/or apply it in one command. But this `tf-all` is a function that we implemented that you can find here https://github.com/geoadmin/infra-ansible-bgdi-collection

Reply:
Awesome. Should we call this out in this documentation? Or even better if we can make the CI show the output of "tf-all plan" on every PR :)

Reply:
For the CI I would not use tf-all, there is much better tool like atlantis for this, so if we want CI to do more than just checking formatting I would first consider such tool before trying to implement something ourself.
I already thought moving the `tf-all` command outside the infra-ansible-bgdi-collection to make it public and provide it as a oh-my-zsh plugin, this way the community could also use it and we could mention it here, but this would require some work.